src/commons/api/rrdp.rs

```diff
@@ -313,6 +316,29 @@ impl CurrentObjects {
         }
     }
 
+    /// Returns a copy of self where elements matching the given URI
+    /// are removed if there are any matches. Otherwise, returns None.
+    pub fn purge(&self, uri: &uri::Rsync) -> Option<Self> {
+        let mut withdraws = vec![];
+
+        // We first loop through the elements to avoid having to clone in case there is no work
+        for (hash, el) in &self.0 {
+            if el.uri() == uri || (uri.as_str().ends_with('/') && el.uri().as_str().starts_with(uri.as_str())) {
+                withdraws.push(hash)
+            }
+        }
+
+        if withdraws.is_empty() {
+            None
+        } else {
+            let mut copy_of_self = self.clone();
```

Comment:
You could also construct the `HashMap` using `FromIterator` on only those items matching a filter, instead of clone then remove.

Reply:
yeah, but then we get a nested loop - i.e. O-square (well N el *M filter) rather than O-N (I presume) for clone, and O log M I presume for remove. Mostly we will be removing only a a fraction - so I think the allocation and then removal should be fine.